### PR TITLE
Count qualifier to view n issues in the list view #1140

### LINKED ISSUE #1140: Count qualifier to view n issues in the list view
For example, there should be a way to get the 10 issues that was most recently updated.

### PR DESCRIPTION
**Pull request by codemanmeet**
Fixes #1140.

### CODE REVIEW

**Comment by damithc** [comment]
`sort` should be applied before `count` if both are present, am I right?.

**Comment by codemanmeet** [comment]
Yup, sort takes precedence that way! `count` is applied to the final issue list that is to be displayed in the list view.

File: src/main/java/filter/expression/Qualifier.java [hunk 1]
```diff
@@ -243,6 +243,8 @@ public boolean isSatisfiedBy(IModel model, TurboIssue issue, MetaQualifierInfo i
             return satisfiesUpdatedHours(issue);
         case REPO:
             return satisfiesRepo(issue);
+        case COUNT:
+                return satisfiesCount();
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: `count` is a meta-qualifier (like `in` and `sort`), so it should be [stripped](https://github.com/HubTurbo/HubTurbo/blob/master/src/main/java/filter/expression/Qualifier.java#L405) by the time we get here. This method shouldn't have to be changed.

File: src/main/java/ui/listpanel/ListPanel.java [hunk 2]
```diff
@@ -301,6 +303,22 @@ private ContextMenu updateContextMenu(ContextMenu contextMenu) {
         return contextMenu;
     }
 
+    private ObservableList<TurboIssue> applyCountQualifierToIssueList(ObservableList<TurboIssue> issueList,
```
Comment (dariusf) [inline src/main/java/ui/listpanel/ListPanel.java]: I'm not sure this is the right place to put this -- we want to try and keep logic out of the UI. By the time the issues get to the UI to be shown, all the effects of filters should already have been applied. You might want to look into how `sort` is implemented, and do something similar for this.

File: src/main/java/filter/expression/Qualifier.java (hunk 1 above)
Comment (codemanmeet) [inline src/main/java/filter/expression/Qualifier.java]: Sure, will move it to the meta qualifier list and add it to the `stripped` method!

File: src/main/java/ui/listpanel/ListPanel.java (hunk 2 above)
Comment (codemanmeet) [inline src/main/java/ui/listpanel/ListPanel.java]: Yup, looked at `sort`. I have added two methods in `UpdateController` that do the job of applying the count filter. `sort` is given preference over `count`.

**Comment by codemanmeet** [comment]
@dariusf help me review it again when you are free!

File: src/main/java/backend/Logic.java [hunk 3]
```diff
@@ -218,7 +218,7 @@ public void replaceIssueLabelsUI(TurboIssue issue, List<String> labels) {
      * and then sends the data to the GUI.
      */
     private void refreshUI() {
-        updateController.filterSortRefresh(getAllUIFilters());
+        updateController.filterSortCountRefresh(getAllUIFilters());
```
Comment (dariusf) [inline src/main/java/backend/Logic.java]: Hmm, I don't think we need to change this method name -- `count` is kind of like a filter. It can't keep growing as we add more features.

File: src/main/java/backend/UpdateController.java [hunk 4]
```diff
@@ -50,10 +51,10 @@ public void filterSortRefresh(List<FilterExpression> filterExprs) {
                             + results.size() + " repos"))
                     .thenCompose(n -> logic.getRateLimitResetTime())
                     .thenApply(logic::updateRemainingRate)
-                    .thenRun(() -> logic.updateUI(filterAndSort(filterExprs))); // Then filter the second time.
+                    .thenRun(() -> logic.updateUI(filterSortAndCount(filterExprs))); // Then filter the second time.
```
Comment (dariusf) [inline src/main/java/backend/UpdateController.java]: Maybe use a more general name, like `processFilter`? `withFilter`?

File: src/main/java/backend/UpdateController.java [hunk 5]
```diff
@@ -112,11 +113,12 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
                         .sorted(determineComparator(filterExprNoAlias, hasUpdatedQualifier))
                         .collect(Collectors.toList());
 
-                filteredAndSorted.put(filterExpr, filteredAndSortedIssues);
+                filteredSortedAndCounted.put(filterExpr,
```
Comment (dariusf) [inline src/main/java/backend/UpdateController.java]: Instead of doing this, you can add `.limit(n)` after `.sorted(...)`. That will take care of truncating the stream so only n or fewer elements remain after the entire thing is collected.

File: src/main/java/backend/UpdateController.java [hunk 6]
```diff
@@ -144,4 +146,40 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
...
     }
+
+    /**
+     * Applies the count qualifier to the set of issues obtained after the sorting order is applied.
+     *
+     * @param issueList Issue list obtained after filtering and sorting.
+     * @param expression The filter expression of the particular panel.
+     * @return The modified issue list with the count applied.
+     */
+    private List<TurboIssue> applyCountQualifierToIssueList(List<TurboIssue> issueList, FilterExpression expression){
```
Comment (dariusf) [inline src/main/java/backend/UpdateController.java]: These two functions can be combined into one that just returns the number of elements to `limit` the stream to.

Also, instead of throwing an exception, maybe just use the first one, like `sort` does?

File: src/main/java/filter/expression/Qualifier.java [hunk 7]
```diff
@@ -441,6 +443,15 @@ public static boolean isUpdatedQualifier(Qualifier q) {
         }
     }
 
+    public static boolean isCountQualifier(Qualifier q) {
+        switch (q.getType()) {
```
Comment (dariusf) [inline src/main/java/filter/expression/Qualifier.java]: This can just be `q.getType() == QualifierType.COUNT`. The use of switch is left over from when the type was a string and we had to account for aliases.

File: src/main/java/backend/UpdateController.java (hunk 6 above)
Comment (codemanmeet) [inline src/main/java/backend/UpdateController.java]: Sure will change to one function and will use the first one

File: src/main/java/filter/expression/Qualifier.java (hunk 7 above)
Comment (codemanmeet) [inline src/main/java/filter/expression/Qualifier.java]: Yup, will change it

File: src/main/java/backend/UpdateController.java (hunk 5 above)
Comment (codemanmeet) [inline src/main/java/backend/UpdateController.java]: Using `limit` now!

File: src/main/java/backend/UpdateController.java (hunk 4 above)
Comment (codemanmeet) [inline src/main/java/backend/UpdateController.java]: changed it to `processFilter`.

File: src/main/java/backend/UpdateController.java [hunk 8]
```diff
@@ -144,4 +145,22 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
...
+     * Applies the count qualifier to the set of issues obtained after the sorting order is applied.
+     *
+     * @param issueList Issue list obtained after filtering and sorting.
+     * @param filterExpr The filter expression of the particular panel.
+     * @return The maximum count of issues
+     */
+    private int determineMaxCount(List<TurboIssue> issueList, FilterExpression filterExpr){
+        for (Qualifier metaQualifier : filterExpr.find(Qualifier::isMetaQualifier)) {
+            if (metaQualifier.getType() == QualifierType.COUNT) {
+                int countQualifierNum = metaQualifier.getNumber().get();
```
Comment (dariusf) [inline src/main/java/backend/UpdateController.java]: This would fail for an input like `count:a`, because then `getNumber()` would be empty. You should probably check for the presence of `getNumber()` in the condition. Unfortunately we don't have a way to statically rule out invalid inputs yet.

File: src/main/java/backend/UpdateController.java [hunk 9]
```diff
@@ -144,4 +145,22 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
...
+                return countQualifierNum < issueList.size() && countQualifierNum >= 0 ?
```
Comment (dariusf) [inline src/main/java/backend/UpdateController.java]: `limit` also handles this comparison for us. You can actually just return `countQualifierNum` here. The 'max' can be removed from the method name to reflect that.

**Comment by codemanmeet** [comment]
Let me know if it can be merged!

File: src/main/java/backend/UpdateController.java [hunk 10]
```diff
@@ -144,4 +145,20 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
...
+     * Applies the count qualifier to the set of issues obtained after the sorting order is applied.
+     *
+     * @param issueList Issue list obtained after filtering and sorting.
+     * @param filterExpr The filter expression of the particular panel.
+     * @return The count value in the qualifier
+     */
+    private int determineCount(List<TurboIssue> issueList, FilterExpression filterExpr){
+        for (Qualifier metaQualifier : filterExpr.find(Qualifier::isMetaQualifier)) {
+            if (metaQualifier.getType() == QualifierType.COUNT) {
+                return metaQualifier.getNumber().isPresent() ? metaQualifier.getNumber().get() : issueList.size();
```
Comment (dariusf) [inline src/main/java/backend/UpdateController.java]: What I meant was to have the `isPresent` check in the if condition, so if the user input is

```
count:a count:1 count:2
```
it shows one issue, i.e. it takes the first *valid* qualifier, not the first one regardless of validity.

Right now it would see `count:a` then default to the size of the list.

You might want to have a test that has multiple qualifiers also, to demonstrate that this is what is intended.

**Comment by dariusf** [comment]
Sorry, this was quite drawn out. Last round.

**Comment by codemanmeet** [comment]
Oh, I didn't know I had to choose the first *valid* qualifier. I was choosing the first qualifier itself. I'll change that.

**Comment by damithc** [comment]
Long drawn out reviews are a sign of high expectations in code quality. I'm happy to see those. :-p

**Comment by codemanmeet** [comment]
Haha true true :-p

**Comment by dariusf** [comment]
Okay, this looks good now. Thanks for making all the changes. Could you update to the latest master before merging?

File: src/test/java/guitests/FilterTests.java [hunk 11]
```diff
@@ -64,6 +64,91 @@ public void milestoneAliasFilterTest() {
...
+
+        assertEquals(2, issuePanel.getIssueCount());
+
+        // Checking 7 issues shown for count:7
+        click("#dummy/dummy_col0_filterTextField");
+        selectAll();
+        type("count:7");
+        push(KeyCode.ENTER);
+
+        assertEquals(7, issuePanel.getIssueCount());
```
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: Why 2 and 7? Are they boundary values?

File: src/test/java/guitests/FilterTests.java [hunk 12]
```diff
@@ -64,6 +64,91 @@ public void milestoneAliasFilterTest() {
...
+
+        //first count qualifier chosen
+        click("#dummy/dummy_col0_filterTextField");
+        selectAll();
+        type("count:6 count:9");
+        push(KeyCode.ENTER);
+
+        assertEquals(6, issuePanel.getIssueCount());
+
+        // Not-a-number returns the whole list
```
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: 'not a number' should return nothing instead?

File: src/test/java/guitests/FilterTests.java [hunk 13]
```diff
@@ -64,6 +64,91 @@ public void milestoneAliasFilterTest() {
...
+        type("count:abcd");
+        push(KeyCode.ENTER);
+
+        assertEquals(10, issuePanel.getIssueCount());
+
+        // Return the value from the first valid count qualifier
+
+        click("#dummy/dummy_col0_filterTextField");
+        selectAll();
+        type("count:abc count:8 count:2");
```
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: There is an implicit AND between clauses, so should return the smallest number because count is interpreted as 'at most' ? i.e. 'at most 2 AND at most 5' should equal to 'at most 2' right?

File: src/test/java/guitests/FilterTests.java [hunk 14]
```diff
@@ -64,6 +64,91 @@ public void milestoneAliasFilterTest() {
...
+        assertEquals(8, issuePanel.getIssueCount());
+        assertEquals("Issue 10", issuePanel.getIssueList().get(0).getTitle());
+        assertEquals("Issue 9", issuePanel.getIssueList().get(1).getTitle());
+
+        // Test with sort qualifier
+
+        click("#dummy/dummy_col0_filterTextField");
+        selectAll();
+        type("count:8 count:2 sort:unmerged");
+        push(KeyCode.ENTER);
```
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: Why 2 count clauses and sort are tested together?

File: src/test/java/guitests/FilterTests.java [hunk 15]
```diff
@@ -64,6 +64,91 @@ public void milestoneAliasFilterTest() {
...
+
+        assertEquals(8, issuePanel.getIssueCount());
+        assertEquals("Issue 1", issuePanel.getIssueList().get(0).getTitle());
+        assertEquals("Issue 2", issuePanel.getIssueList().get(1).getTitle());
+    }
+
```
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: @dariusf can filters be tested without involving GUI?

File: src/main/java/backend/UpdateController.java [hunk 16]
```diff
@@ -144,4 +145,20 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
...
     }
+
+    /**
+     * Applies the count qualifier to the set of issues obtained after the sorting order is applied.
+     *
+     * @param issueList Issue list obtained after filtering and sorting.
+     * @param filterExpr The filter expression of the particular panel.
+     * @return The count value in the qualifier
+     */
+    private int determineCount(List<TurboIssue> issueList, FilterExpression filterExpr){
```
Comment (damithc) [inline src/main/java/backend/UpdateController.java]: The comment says 'Applies' and the method name says 'determines'. Which is it?

File: src/main/java/backend/UpdateController.java [hunk 17]
```diff
@@ -93,30 +93,31 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
      * @param filterExprs Filter expressions to process.
      * @return Filter expressions and their corresponding issues after filtering and sorting.
      */
```
Comment (damithc) [inline src/main/java/backend/UpdateController.java]: This header comment needs updating?

File: src/main/java/backend/UpdateController.java [hunk 18]
```diff
@@ -93,30 +93,31 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
...
 
             if (filterAndSortedExpression == null) { // If it already exists, no need to filter anymore
                 boolean hasUpdatedQualifier = Qualifier.hasUpdatedQualifier(filterExpr);
 
                 FilterExpression filterExprNoAlias = Qualifier.replaceMilestoneAliases(models, filterExpr);
 
-                List<TurboIssue> filteredAndSortedIssues = allModelIssues.stream()
+                List<TurboIssue> filteredSortedAndCountedIssues = allModelIssues.stream()
                         .filter(issue -> Qualifier.process(models, filterExprNoAlias, issue))
                         .sorted(determineComparator(filterExprNoAlias, hasUpdatedQualifier))
```
Comment (damithc) [inline src/main/java/backend/UpdateController.java]: not related to this issue, but shouldn't this be `sort`, not `sorted`?

File: src/main/java/backend/UpdateController.java [hunk 19]
```diff
@@ -144,4 +145,20 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
         // No sort or updated, return sort by descending ID, which is the default.
         return Qualifier.getSortComparator(models, "id", true, false);
     }
+
+    /**
+     * Applies the count qualifier to the set of issues obtained after the sorting order is applied.
```
Comment (damithc) [inline src/main/java/backend/UpdateController.java]: This method has nothing to do with sorting. It should not be mentioned in the header comment. If this was a proper description of the method, I should see something like ' ... else returns the size of the issueList'

File: src/test/java/guitests/FilterTests.java (hunk 15 above)
Comment (dariusf) [inline src/test/java/guitests/FilterTests.java]: Yes for the true/false ones, but no for `sort` and `count`. This will require some refactoring. Should we do it now?

File: src/test/java/guitests/FilterTests.java (hunk 13 above)
Comment (dariusf) [inline src/test/java/guitests/FilterTests.java]: Okay, this may be a better way to interpret it. Right now we're just doing the simplest thing and taking the first (leftmost, if we think of the expression as a tree) valid qualifier. Same for `sort` if there are multiple qualifiers.

File: src/test/java/guitests/FilterTests.java (hunk 15 above)
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: I see. May be as a separate issue then. At the same time need to move some of the testing done in gui to lower level unit tests. Otherwise, future devs follow existing code and dump all testing into GUI tests.

File: src/main/java/backend/UpdateController.java (hunk 18 above)
Comment (dariusf) [inline src/main/java/backend/UpdateController.java]: That's a method from the new streams API. I think the convention is that verbs work on one element at a time, but adjectives work on the stream as a whole.

File: src/test/java/guitests/FilterTests.java (hunk 13 above)
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: I see. First, we need to document it as users might expect a different behavior based on their intuition. Second, let's create issues to make the behavior more intuitive or make the limitation more explicit e.g. flag multiple sorts/counts as syntax errors.

File: src/main/java/backend/UpdateController.java (hunk 18 above)
Comment (damithc) [inline src/main/java/backend/UpdateController.java]: Got it.

File: src/test/java/guitests/FilterTests.java (hunk 11 above)
Comment (codemanmeet) [inline src/test/java/guitests/FilterTests.java]: Nope, 0 and 10 are boundary values. Both have been tested. In addition, 2 and 7 are being tested as well.

File: src/test/java/guitests/FilterTests.java (hunk 14 above)
Comment (codemanmeet) [inline src/test/java/guitests/FilterTests.java]: They are being tested to check for sort's precedence over count.

File: src/test/java/guitests/FilterTests.java (hunk 12 above)
Comment (codemanmeet) [inline src/test/java/guitests/FilterTests.java]: For all qualifiers. `updated`, `sort` etc. if you enter the wrong data type (`string` instead of `number`), the entire list is returned.

File: src/main/java/backend/UpdateController.java (hunk 17 above)
Comment (codemanmeet) [inline src/main/java/backend/UpdateController.java]: Yup, will update the header comment.

File: src/main/java/backend/UpdateController.java (hunk 16 above)
Comment (codemanmeet) [inline src/main/java/backend/UpdateController.java]: It is `determines`. Will change it to that.

File: src/test/java/guitests/FilterTests.java (hunk 12 above)
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: Should we reverse that? Otherwise it is misleading. @dariusf what do you think?

File: src/test/java/guitests/FilterTests.java (hunk 14 above)
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: Having 2 count clauses is not necessary for that right? Instead, may be have test cases where sort comes before count in the filter, and vice versa.

File: src/test/java/guitests/FilterTests.java (hunk 11 above)
Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: Just one non boundary case is enough? If 2 is broken, 7 has to be broken as well right?

Comment (damithc) [inline src/test/java/guitests/FilterTests.java]: Did you test for a negative number?

Comment (codemanmeet) [inline src/test/java/guitests/FilterTests.java]: I added a test for negative numbers as well.

File: src/test/java/guitests/FilterTests.java (hunk 12 above)
Comment (codemanmeet) [inline src/test/java/guitests/FilterTests.java]: Sorry, my bad. For `updated`, if you enter the wrong data type (string), empty list is returned. If all count qualifiers entered are invalid (for eg.), shall I return an empty list?

File: src/main/java/backend/UpdateController.java [hunk 20]
```diff
@@ -144,4 +145,20 @@ private void openRepositoriesInFilters(List<FilterExpression> filterExprs) {
...
+     * @param filterExpr The filter expression of the particular panel.
+     * @return The count value in the qualifier
+     */
+    private int determineCount(List<TurboIssue> issueList, FilterExpression filterExpr){
+        for (Qualifier metaQualifier : filterExpr.find(Qualifier::isMetaQualifier)) {
+            if (metaQualifier.getType() == QualifierType.COUNT && metaQualifier.getNumber().isPresent()) {
+                return metaQualifier.getNumber().get();
+            }
+        }
+        return issueList.size();
```
Comment (damithc) [inline src/main/java/backend/UpdateController.java]: This line doesn't match with the header comment.

Comment (codemanmeet) [inline src/main/java/backend/UpdateController.java]: Changed it.

**Comment by codemanmeet** [comment]
@damithc As discussed on Slack, I'll go ahead and implement `Parse Exception` for any invalid `count` qualifier found.

**Comment by dariusf** [comment]
Note that on a parse exception, we show the entire list.

So in general, should we always throw exceptions on invalid inputs? A lot of qualifiers just default to empty (return `true` on everything) if there is something unexpected.

Also, any changes to docs needed for this?

**Comment by codemanmeet** [comment]
Yup, that's how we want it. `Parse Exception` returns `Qualifier.EMPTY` and shows the entire list. The exception informs the user of the error at the bottom status bar of the pView. He/she should still see the issues and not be taken aback by an empty list. The user can then type in a fresh filter and then rectify the error.

**Comment by damithc** [comment]
I'm inclined to show an error upon syntax errors. If we do show the full list on such cases, there should be a visual indication that the list is show due to an error, not as the filter result. Otherwise, we can show an empty list, but again we should indicate the empty list is due to the error, not because nothing matches the filter.
The alternative is to ignore incorrect part of the syntax. Is there any benefit of doing that? The downside is the user might be mislead.

**Comment by dariusf** [comment]
Yes, I guess there's not much point in interpreting it wrongly.

**Comment by codemanmeet** [comment]
@dariusf `Parse exception` has been added for `count` with relevant error messages. Commits have been squashed. Thanks for the review and creating the pertinent issues! :) As discussed, will go ahead and merge once the CI passes.
